Implement KZG proof batch verification (option 2) - uses worker pool

terencechain opened the pull request:
This is an alt to #15609, it changes KZG proof batch verification from timer-based batching to a responsive worker pool model that processes messages immediately when idle and buffers when busy

terencechain committed:
Implement KZG proof batch verification for data column gossip validation

nalepae commented on beacon-chain/sync/batch_verifier.go:
```diff
@@ -2,22 +2,40 @@ package sync
 )
 
-const signatureVerificationInterval = 50 * time.Millisecond
+const (
```
Nit: Why changing this?

nalepae commented on beacon-chain/sync/validate_data_column.go:
```diff
@@ -145,9 +145,12 @@ func (s *Service) validateDataColumn(ctx context.Context, pid peer.ID, msg *pubs
 	// [REJECT] The sidecar's column data is valid as verified by `verify_data_column_sidecar_kzg_proofs(sidecar)`.
-	if err := verifier.SidecarKzgProofVerified(); err != nil {
-		return pubsub.ValidationReject, err
+	validationResult, err := s.validateWithKzgBatchVerifier(ctx, "data column KZG proof", roDataColumns)
```
Nit: Should we really pass `"data column KZG proof"` as an argument? This function can only take RO data column sidecars as input. So I guess we can hard code `message` in the function itself if needed.

nalepae commented on beacon-chain/sync/batch_verifier.go:
```diff
@@ -120,3 +213,47 @@ func performBatchAggregation(aggSet *bls.SignatureBatch) (*bls.SignatureBatch, e
+		return
+	}
+
+	allDataColumns := make([]blocks.RODataColumn, 0)
```
```suggestion
	allDataColumns := make([]blocks.RODataColumn, 0, len(kzgBatch))
```

nalepae commented on beacon-chain/sync/batch_verifier.go:
```diff
@@ -47,6 +65,81 @@ func (s *Service) verifierRoutine() {
+		// Start immediate processing
+		pool.processing = true
+		pool.currentBatch = []*kzgVerifier{kzg}
+		go pool.processBatch()
```
Alternative design proposal:
I had the same issue when wanting to save data column sidecars into the file system. I wanted to be sure that, before closing the file, no other pending sidecars were to save (because they were added to the queue during the current save process).

I used that:
https://github.com/OffchainLabs/prysm/blob/1f07a5c2b1643a7b40bb4f4b290db0257d6084b1/beacon-chain/db/filesystem/data_column.go#L965-L978.
This way, I'm able to iterate over the channel until it is totally empty and process all pending sidecars in the same batch, without having to use an extra go-routine.
In this case, `kzgVerifierRoutine` could use this system and `processMessage` could be blocking. At the next iteration of the for loop in `kzgVerifierRoutine`, it will totally empty the values stored in the chan, and process them. Please ping me if I'm unclear!

nalepae commented on beacon-chain/sync/batch_verifier.go:
```diff
@@ -120,3 +213,47 @@ func performBatchAggregation(aggSet *bls.SignatureBatch) (*bls.SignatureBatch, e
+	close(resChan)
+	if resErr != nil {
+		log.WithError(resErr).Tracef("Could not perform batch verification of %s", message)
+		err := peerdas.VerifyDataColumnsSidecarKZGProofs(dataColumns)
```
I think a comment indicating why we re-execute `VerifyDataColumnsSidecarKZGProofs` in this case could be useful.

nalepae commented on beacon-chain/sync/service.go:
```diff
@@ -202,6 +203,8 @@ func NewService(ctx context.Context, opts ...Option) *Service {
 	// Initialize signature channel with configured limit
 	r.signatureChan = make(chan *signatureVerifier, r.cfg.batchVerifierLimit)
+	// Initialize KZG channel for worker pool
+	r.kzgChan = make(chan *kzgVerifier, 100)
```
I think adding a little rationale explaining how this `100` has been designed could be useful.

nalepae reviewed (commented):
I did not run the code itself.

terencechain committed:
Manu's feedback

terencechain committed:
Add tests

satushh commented on beacon-chain/sync/batch_verifier.go:
```diff
@@ -120,3 +171,51 @@ func performBatchAggregation(aggSet *bls.SignatureBatch) (*bls.SignatureBatch, e
+	verificationSet := &kzgVerifier{dataColumns: dataColumns, resChan: resChan}
+	s.kzgChan <- verificationSet
+
+	resErr := <-resChan
```
Should this be wrapped in a select statement with ctx.Done()

kasey replied:
I think so too, and actually I would suggest using `context.WithTimeout` here to limit how long this potentially blocks on batch verification. This would simplify the other side of the channel because it would no longer need to write ctx.Err() back to all the channels (which racy anyway and can leave these goroutines wedged).

nalepae committed:
Merge branch 'develop' into verify-kzg-proofs1

nalepae commented on beacon-chain/sync/batch_verifier.go:
```diff
@@ -120,3 +171,51 @@ func performBatchAggregation(aggSet *bls.SignatureBatch) (*bls.SignatureBatch, e
+	for i := 0; i < len(kzgBatch); i++ {
+		kzgBatch[i].resChan <- verificationErr
+	}
+}
```
Nit

```suggestion
	for _, verifier := range kzgBatch {
		verifier.resChan <- verificationErr
	}
}
```

nalepae commented on beacon-chain/sync/validate_data_column.go:
```diff
@@ -145,9 +145,12 @@ func (s *Service) validateDataColumn(ctx context.Context, pid peer.ID, msg *pubs
+		return validationResult, err
 	}
+	// Mark KZG verification as satisfied since we did it via batch verifier
+	verifier.SatisfyRequirement(verification.RequireSidecarKzgProofVerified)
```
Why not using
```go
defer dv.recordResult(RequireSidecarKzgProofVerified, &err)
```

directly in `s.validateWithKzgBatchVerifier` as done for all other verification functions?

nalepae commented on beacon-chain/sync/kzg_batch_verifier_test.go:
```diff
@@ -0,0 +1,365 @@
+		const numRequests = 5
+		resChans := make([]chan error, numRequests)
+
+		for i := 0; i < numRequests; i++ {
```
Nit

```suggestion
		for i := range numRequests {
```

nalepae commented on beacon-chain/sync/kzg_batch_verifier_test.go:
```diff
@@ -0,0 +1,365 @@
+			service.kzgChan <- &kzgVerifier{dataColumns: dataColumns, resChan: resChan}
+		}
+
+		for i := 0; i < numRequests; i++ {
```
```suggestion
		for i := range numRequests {
```

nalepae commented on beacon-chain/sync/kzg_batch_verifier_test.go:
```diff
@@ -0,0 +1,365 @@
+		go func(goroutineID int) {
+			defer wg.Done()
+
+			for j := 0; j < numRequestsPerGoroutine; j++ {
```
Nit

```suggestion
			for j := range numRequestsPerGoroutine {
```

nalepae commented on beacon-chain/sync/kzg_batch_verifier_test.go:
```diff
@@ -0,0 +1,365 @@
+		}
+
+		const numVerifications = 3
+		for i := 0; i < numVerifications; i++ {
```
Nit

```suggestion
		for i := range numVerifications {
```

nalepae commented on beacon-chain/sync/kzg_batch_verifier_test.go:
```diff
@@ -0,0 +1,365 @@
+	wg.Add(numGoroutines)
+
+	// Multiple goroutines sending verification requests simultaneously
+	for i := 0; i < numGoroutines; i++ {
```
Nit


```suggestion
	for i := range numGoroutines {
```

terencechain committed:
Update beacon-chain/sync/batch_verifier.go

Co-authored-by: Manu NALEPA <enalepa@offchainlabs.com>

terencechain committed:
Update beacon-chain/sync/batch_verifier.go

Co-authored-by: Manu NALEPA <enalepa@offchainlabs.com>

terencechain committed:
Update beacon-chain/sync/kzg_batch_verifier_test.go

Co-authored-by: Manu NALEPA <enalepa@offchainlabs.com>

terencechain committed:
Update beacon-chain/sync/kzg_batch_verifier_test.go

Co-authored-by: Manu NALEPA <enalepa@offchainlabs.com>

terencechain committed:
Update beacon-chain/sync/kzg_batch_verifier_test.go

Co-authored-by: Manu NALEPA <enalepa@offchainlabs.com>

terencechain committed:
Update beacon-chain/sync/kzg_batch_verifier_test.go

Co-authored-by: Manu NALEPA <enalepa@offchainlabs.com>

terencechain committed:
Update beacon-chain/sync/kzg_batch_verifier_test.go

Co-authored-by: Manu NALEPA <enalepa@offchainlabs.com>

terencechain committed:
Fix tests

kasey commented on beacon-chain/sync/batch_verifier.go:
```diff
@@ -47,6 +54,50 @@ func (s *Service) verifierRoutine() {
+			verifyKzgBatch(kzgBatch)
+		}
+	}
+}
```
I think you could simplify the pair of methods `kzgVerifierRoutine` + `pullKzgChan` into a single method like the following. Also note that I have dropped this `kzg.resChan <- s.ctx.Err()` loop - I think this is not necessary if the callers use a select and a context.WithTimeout (which they should do anyway for safe concurrency).

```
// A routine that runs in the background to perform batch
// KZG verifications by draining the channel and processing all pending requests.
func (s *Service) kzgVerifierRoutine() {
    kzgBatch := make([]*kzgVerifier, 0, 1)
    for {
		select {
			case <-s.ctx.Done():
				return
			case kzg := <-s.kzgChan:
				kzgBatch = append(kzgBatch, kzg)
				continue
			default:
				verifyKzgBatch(kzgBatch)
				kzgBatch = make([]*kzgVerifier, 0, 1)
		}
	}
}
```
(edit: fixed a bug in the suggestion)

kasey commented on beacon-chain/sync/batch_verifier.go:
```diff
@@ -120,3 +171,51 @@ func performBatchAggregation(aggSet *bls.SignatureBatch) (*bls.SignatureBatch, e
+	s.kzgChan <- verificationSet
+
+	resErr := <-resChan
+	close(resChan)
```
You don't really need to close channels, they get cleaned up by the garbage collector. You should only close a channel on the sending side, to unblock any receivers who could be waiting on more values from a sender which isn't going to produce any more. Receiving from a closed channel is fine (instantly unblocks any goroutine trying to receive), but sending to one /causes a panic/. This feedback will be more critical if you take the other piece of feedback to check `ctx.Done()` above, because in that instance you won't have read from `resChan`, so the other side could easily send to it after it is closed here and panic.

kasey commented on beacon-chain/sync/batch_verifier.go:
```diff
@@ -120,3 +171,51 @@ func performBatchAggregation(aggSet *bls.SignatureBatch) (*bls.SignatureBatch, e
+		// the individual data columns might still be valid.
+		err := peerdas.VerifyDataColumnsSidecarKZGProofs(dataColumns)
+		if err != nil {
+			verErr := errors.Wrapf(err, "Could not verify")
```
nitpick: Why a new variable and not just `err = errors.Wrapf(err, "Could not verify")`?

kasey commented on beacon-chain/sync/service.go:
```diff
@@ -254,6 +259,7 @@ func (s *Service) Start() {
 	s.newColumnsVerifier = newDataColumnsVerifierFromInitializer(v)
 
 	go s.verifierRoutine()
+	go s.kzgVerifierRoutine()
```
Should we have more than 1 goroutine doing this?

terencechain committed:
Kasey's feedback

nalepae committed:
`validateWithKzgBatchVerifier`: Give up after a full slot.

Before this commit:
After 100 ms, an un-batched verification is launched concurrently to the batched one.
As a result, a stressed node could start to be even more stressed by the multiple verifications.
Also, it is always hard to choose a correct timeout value.
100ms may be OK for a given node with a given BPO version, and not ok for the same node with a BPO version with 10x more blobs.

However, we know this gossip validation won't be useful after a full slot duration.

After this commit:
After a full slot duration, we just ignore the incoming gossip message.
It's important to ignore it and not to reject it, since rejecting it would downscore the peer sending this message.

terencechain committed:
Merge branch 'develop' into verify-kzg-proofs1
